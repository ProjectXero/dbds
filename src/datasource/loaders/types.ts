import { QueryOptions } from '../queries/QueryBuilder'

export type SearchableKeys<T, SearchableType = string | number | null> = {
  [K in keyof T]?: T extends { [_ in K]?: SearchableType } ? K : never
}[keyof T]

export interface GetDataFunction<TRowType> {
  <TColumnName extends keyof TRowType & string>(
    args: Array<TRowType[TColumnName]> | ReadonlyArray<TRowType[TColumnName]>,
    column: TColumnName,
    type: string,
    options?: QueryOptions<TRowType>
  ): readonly TRowType[] | Promise<readonly TRowType[]>
}

export interface LoaderFactoryOptions<TRowType> {
  columnToKey?: (column: string) => string
  keyToColumn?: (column: string) => string
  columnTypes: Record<keyof TRowType, string>
}

export type LoaderOptions<TRowType, TColumnName extends keyof TRowType> = {
  getData?: GetDataFunction<TRowType>
  multi?: boolean
  ignoreCase?: TRowType extends Record<TColumnName, string>
    ? boolean
    : false | undefined
  callbackFn?: (
    row: TRowType,
    index: number,
    array: readonly TRowType[]
  ) => void
} & QueryOptions<TRowType>

export type FinderOptions = {
  multi?: boolean
}
