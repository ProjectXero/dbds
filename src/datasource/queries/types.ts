import { SqlSqlTokenType } from "slonik"

export type PrimitiveValueType = string | number | boolean | null
export type ValueType =
  | PrimitiveValueType
  | Array<PrimitiveValueType>
  | ReadonlyArray<PrimitiveValueType>

export type ConditionValue<T extends string | number | boolean> = T | null | Array<T | null>

export type GenericConditions = Record<string, ConditionValue<string> | ConditionValue<number> | ConditionValue<boolean> | SqlSqlTokenType>

export type Conditions<TRowType> = {
  [K in keyof TRowType]?: TRowType[K] extends string
  ? ConditionValue<string>
  : TRowType[K] extends number
  ? ConditionValue<number>
  : TRowType[K] extends boolean
  ? ConditionValue<boolean>
  : SqlSqlTokenType
} & GenericConditions
