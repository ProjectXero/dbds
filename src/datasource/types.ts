import { AsyncLocalStorage } from 'async_hooks'
import DataLoader from 'dataloader'
import {
  DatabasePool,
  DatabaseTransactionConnection,
  IdentifierNormalizer,
} from 'slonik'
import { z } from 'zod'

import { QueryOptions as BuilderOptions } from './queries/QueryBuilder'

export interface QueryOptions<Schema, TResultType = Schema>
  extends BuilderOptions<Schema> {
  eachResult?: LoaderCallback<TResultType>
  expected?: 'one' | 'many' | 'maybeOne' | 'any'
  defaultSymbol?: symbol
}

export interface KeyNormalizers {
  keyToColumn: IdentifierNormalizer
  columnToKey: IdentifierNormalizer
}
export type LoaderCallback<TResultType> = (
  value: TResultType,
  index: number,
  array: readonly TResultType[]
) => void

interface AsyncStorage<Schema> {
  transaction: DatabaseTransactionConnection
  loaderLookups: Array<[Loader<Schema>, readonly unknown[]]>
}

type Loader<Schema> = DataLoader<unknown, Schema[] | (Schema | undefined)>

export interface ExtendedDatabasePool<Schema> extends DatabasePool {
  async: AsyncLocalStorage<AsyncStorage<Schema>>
}

export type TableSchema<TableColumns extends string> = z.ZodObject<{
  [K in string & TableColumns]: z.ZodTypeAny
}>
export type ColumnMetadata = {
  nativeName: string
  nativeType: string
}
export type TableMetadata<ColumnNames extends string = string> = {
  [K in string & ColumnNames]: ColumnMetadata
}

export type TableInfo<ColumnNames extends string = string> = {
  name: string
  metadata: TableMetadata<ColumnNames>
  schemas: {
    select: TableSchema<ColumnNames>
    insert: TableSchema<ColumnNames>
    update: TableSchema<ColumnNames>
  }
}
