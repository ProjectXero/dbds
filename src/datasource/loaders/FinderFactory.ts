import DataLoader from 'dataloader'

import { FinderOptions, ExtendedDataLoader } from './types'

export interface FinderFunction<Input, Result> {
  (value: Input): Promise<Result>
}

export const isExtendedDataLoader = <M extends boolean, K, V, C>(
  loader: unknown
): loader is ExtendedDataLoader<M, K, V, C> => {
  return (
    typeof loader === 'object' &&
    typeof (loader as ExtendedDataLoader<M, K, V, C>).isMultiLoader ===
      'boolean'
  )
}

export default class FinderFactory<Schema> {
  public create<TInput>(
    loader: ExtendedDataLoader<true, TInput, Schema[]>,
    options?: FinderOptions
  ): FinderFunction<TInput, Schema[]>
  public create<TInput>(
    loader: ExtendedDataLoader<false, TInput, Schema | undefined>,
    options?: FinderOptions & { multi?: false }
  ): FinderFunction<TInput, Schema | null>
  public create<TInput>(
    loader: DataLoader<TInput, Schema[]>,
    options: FinderOptions & { multi: true }
  ): FinderFunction<TInput, Schema[]>
  public create<TInput>(
    loader: DataLoader<TInput, Schema | undefined>,
    options?: FinderOptions & { multi?: false }
  ): FinderFunction<TInput, Schema | null>
  public create<TInput>(
    loader:
      | ExtendedDataLoader<boolean, TInput, Schema | Schema[] | undefined>
      | DataLoader<TInput, Schema | Schema[] | undefined>,
    options: FinderOptions = {}
  ): FinderFunction<TInput, Schema[] | Schema | null> {
    if (isExtendedDataLoader(loader)) {
      options.multi = loader.isMultiLoader
    }

    const { multi = false } = options

    if (multi === true) {
      return async (value: TInput): Promise<Schema[]> => {
        const result = await loader.load(value)
        if (!result) {
          return []
        }
        if (Array.isArray(result)) {
          return result
        }
        return [result]
      }
    }

    return async (value: TInput): Promise<Schema | null> => {
      const result = await loader.load(value)
      if (!result) {
        return null
      }
      if (Array.isArray(result)) {
        return result[0]
      }
      return result
    }
  }
}
