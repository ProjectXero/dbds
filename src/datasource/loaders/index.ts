export { default as LoaderFactory } from './LoaderFactory'
export { default as FinderFactory } from './FinderFactory'
