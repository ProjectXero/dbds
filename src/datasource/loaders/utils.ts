export const match = (
  left: unknown,
  right: unknown,
  ignoreCase = false
): boolean => {
  if (typeof left === 'string' && typeof right === 'string' && ignoreCase) {
    return left.localeCompare(right, undefined, { sensitivity: 'accent' }) === 0
  }
  return left === right
}

export const identity = <T>(value: T): T => value
