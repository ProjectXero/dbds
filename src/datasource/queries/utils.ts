import {
  IdentifierSqlToken,
  FragmentSqlToken,
  isSqlToken as isSlonikToken,
  SqlToken,
} from 'slonik'
import { FragmentToken } from 'slonik/dist/src/tokens'
import { OrderTuple } from './types'

export const isSqlToken = (subject: unknown): subject is SqlToken => {
  return isSlonikToken(subject)
}

export const isFragmentSqlToken = (v: unknown): v is FragmentSqlToken => {
  return (
    typeof v === 'object' &&
    Object.prototype.hasOwnProperty.call(v, 'sql') &&
    Object.prototype.hasOwnProperty.call(v, 'type') &&
    Object.prototype.hasOwnProperty.call(v, 'values') &&
    (v as FragmentSqlToken).type === FragmentToken
  )
}

export const isIdentifierSqlToken = (v: unknown): v is IdentifierSqlToken => {
  return (
    typeof v === 'object' &&
    Object.prototype.hasOwnProperty.call(v, 'names') &&
    Object.prototype.hasOwnProperty.call(v, 'type')
  )
}

export const isOrderTuple = (v: unknown): v is OrderTuple => {
  return (
    Array.isArray(v) &&
    v.length >= 1 &&
    v.length <= 2 &&
    (v[1] === 'ASC' || v[1] === 'DESC') &&
    (typeof v[0] === 'string' ||
      isIdentifierSqlToken(v[0]) ||
      isFragmentSqlToken(v[0]))
  )
}
