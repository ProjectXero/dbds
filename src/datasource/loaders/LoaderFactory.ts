import DataLoader from 'dataloader'
import { QueryOptions } from '../queries/QueryBuilder'
import { TableMetadata } from '../types'

import {
  ExtendedDataLoader,
  GetDataFunction,
  GetDataMultiFunction,
  LoaderOptions,
  // SearchableKeys,
} from './types'
import { match } from './utils'

export default class LoaderFactory<
  TableSchema,
  SearchableKeys extends string & keyof TableSchema = string & keyof TableSchema
> {
  protected loaders: Array<
    ExtendedDataLoader<
      boolean,
      unknown,
      TableSchema[] | TableSchema | undefined
    >
  > = []

  constructor(
    private getData: GetDataFunction<TableSchema>,
    private getDataMulti: GetDataMultiFunction<TableSchema>,
    private metadata: TableMetadata<SearchableKeys>
  ) {}

  protected async autoPrimeLoaders(
    result: TableSchema | undefined,
    loaders = this.loaders
  ): Promise<void> {
    if (!result) {
      // we can't auto-prime undefined results because we have no values to
      // prime on :(
      return
    }

    for (const loader of loaders) {
      if (loader.isMultiLoader) {
        // auto-priming multi-loaders is currently unsupported
        continue
      }

      if (typeof loader.columns === 'string') {
        loader.prime(result[loader.columns as keyof TableSchema], result)
      } else {
        const key = loader.columns.reduce(
          (key, column) => ({
            ...key,
            [column]: result[column as keyof TableSchema],
          }),
          {}
        )
        loader.prime(key, result)
      }
    }
  }

  public create<TColumnName extends SearchableKeys>(
    key: TColumnName,
    options: LoaderOptions<TableSchema> & {
      multi: true
    },
    queryOptions?: () => QueryOptions<TableSchema>
  ): ExtendedDataLoader<true, TableSchema[TColumnName], TableSchema[]>
  public create<TColumnName extends SearchableKeys>(
    key: TColumnName,
    columnType: string,
    options: LoaderOptions<TableSchema> & {
      multi: true
    },
    queryOptions?: () => QueryOptions<TableSchema>
  ): ExtendedDataLoader<true, TableSchema[TColumnName], TableSchema[]>
  public create<TColumnName extends SearchableKeys>(
    key: TColumnName,
    options?: LoaderOptions<TableSchema> & {
      multi?: false
    },
    queryOptions?: () => QueryOptions<TableSchema>
  ): ExtendedDataLoader<
    false,
    TableSchema[TColumnName],
    TableSchema | undefined
  >
  public create<TColumnName extends SearchableKeys>(
    key: TColumnName,
    columnType?: string,
    options?: LoaderOptions<TableSchema> & {
      multi?: false
    },
    queryOptions?: () => QueryOptions<TableSchema>
  ): ExtendedDataLoader<
    false,
    TableSchema[TColumnName],
    TableSchema | undefined
  >
  public create<
    TColumnName extends SearchableKeys,
    TColType extends TableSchema[TColumnName] & (string | number)
  >(
    key: TColumnName,
    columnType?: string | LoaderOptions<TableSchema>,
    options?: LoaderOptions<TableSchema> | (() => QueryOptions<TableSchema>),
    queryOptions?: () => QueryOptions<TableSchema>
  ): DataLoader<TColType, TableSchema[] | TableSchema | undefined> {
    if (typeof options === 'function') {
      queryOptions = options
      options = undefined
    }
    if (typeof columnType === 'object') {
      options = columnType
      columnType = undefined
    }

    const actualOptions = options || {}

    const getData = actualOptions.getData || this.getData

    const type: string = columnType || this.metadata[key].nativeType

    const {
      multi = false,
      ignoreCase = false,
      callbackFn,
      autoPrime,
      primeLoaders,
    } = actualOptions

    const loader = new DataLoader<
      TColType,
      TableSchema[] | (TableSchema | undefined)
    >(async (args: readonly TColType[]) => {
      const data = await getData<TColumnName>(args, key, type, loader, {
        ...actualOptions,
        ...queryOptions?.(),
      })

      data.forEach((row, idx, arr) => {
        callbackFn && callbackFn(row, idx, arr)
        autoPrime &&
          this.autoPrimeLoaders(
            row,
            typeof primeLoaders === 'function' ? primeLoaders() : primeLoaders
          )
      })

      return args.map((value) => {
        if (multi) {
          return data.filter((row) =>
            match(value, row[key] as TColType, ignoreCase)
          )
        }

        return data.find((row) =>
          match(value, row[key] as TColType, ignoreCase)
        )
      })
    }) as ExtendedDataLoader<
      typeof multi,
      TColType,
      TableSchema[] | (TableSchema | undefined)
    >
    loader.isMultiLoader = multi
    loader.columns = key
    this.loaders.push(loader)
    return loader
  }

  public createMulti<
    TColumnNames extends Array<SearchableKeys>,
    TBatchKey extends {
      [K in TColumnNames[0]]: TableSchema[K]
    }
  >(
    key: TColumnNames,
    options: LoaderOptions<TableSchema, true> & {
      multi: true
    },
    queryOptions?: () => QueryOptions<TableSchema>
  ): ExtendedDataLoader<true, TBatchKey, TableSchema[]>
  public createMulti<
    TColumnNames extends Array<SearchableKeys>,
    TBatchKey extends {
      [K in TColumnNames[0]]: TableSchema[K]
    }
  >(
    key: TColumnNames,
    columnTypes: string[],
    options: LoaderOptions<TableSchema, true> & {
      multi: true
    },
    queryOptions?: () => QueryOptions<TableSchema>
  ): ExtendedDataLoader<true, TBatchKey, TableSchema[]>
  public createMulti<
    TColumnNames extends Array<SearchableKeys>,
    TBatchKey extends {
      [K in TColumnNames[0]]: TableSchema[K]
    }
  >(
    key: TColumnNames,
    options?: LoaderOptions<TableSchema, true> & {
      multi?: false
    },
    queryOptions?: () => QueryOptions<TableSchema>
  ): ExtendedDataLoader<false, TBatchKey, TableSchema | undefined>
  public createMulti<
    TColumnNames extends Array<SearchableKeys>,
    TBatchKey extends {
      [K in TColumnNames[0]]: TableSchema[K]
    }
  >(
    key: TColumnNames,
    columnTypes?: string[],
    options?: LoaderOptions<TableSchema, true> & {
      multi?: false
    },
    queryOptions?: () => QueryOptions<TableSchema>
  ): ExtendedDataLoader<false, TBatchKey, TableSchema | undefined>
  public createMulti<
    TColumnNames extends Array<SearchableKeys>,
    TBatchKey extends {
      [K in TColumnNames[0]]: TableSchema[K]
    }
  >(
    keys: TColumnNames,
    columnTypes?: string[] | LoaderOptions<TableSchema, true>,
    options?:
      | LoaderOptions<TableSchema, true>
      | (() => QueryOptions<TableSchema>),
    queryOptions?: () => QueryOptions<TableSchema>
  ): DataLoader<TBatchKey, TableSchema[] | TableSchema | undefined> {
    if (typeof options === 'function') {
      queryOptions = options
      options = undefined
    }
    if (typeof columnTypes === 'object' && !Array.isArray(columnTypes)) {
      options = columnTypes
      columnTypes = undefined
    }

    const actualOptions = options || {}

    const getData = actualOptions.getData || this.getDataMulti

    const types: string[] =
      columnTypes || keys.map((key) => this.metadata[key].nativeType)

    if (types.length !== keys.length) {
      throw new Error('Same number of types and keys must be provided')
    }

    const cacheKeyFn = (batchKey: TBatchKey): string =>
      keys.map((key) => batchKey[key]).join(':')

    const {
      multi = false,
      ignoreCase = false,
      callbackFn,
      autoPrime,
      primeLoaders,
    } = actualOptions

    const loader = new DataLoader<
      TBatchKey,
      TableSchema[] | (TableSchema | undefined),
      string
    >(
      async (args: readonly TBatchKey[]) => {
        const data = await getData<TColumnNames, TBatchKey>(
          args,
          keys,
          types,
          loader,
          {
            ...actualOptions,
            ...queryOptions?.(),
          }
        )

        data.forEach((row, idx, arr) => {
          callbackFn && callbackFn(row, idx, arr)
          autoPrime &&
            this.autoPrimeLoaders(
              row,
              typeof primeLoaders === 'function' ? primeLoaders() : primeLoaders
            )
        })

        return args.map((batchKey) => {
          const callback = (row: TableSchema) => {
            return Object.entries(batchKey).every(([key, value]) =>
              match(
                value as string | number,
                row[key as keyof TableSchema],
                ignoreCase
              )
            )
          }
          if (multi) {
            return data.filter(callback)
          }

          return data.find(callback)
        })
      },
      { cacheKeyFn }
    ) as ExtendedDataLoader<
      boolean,
      TBatchKey,
      TableSchema[] | (TableSchema | undefined)
    >
    loader.isMultiLoader = multi
    loader.columns = [...keys]
    this.loaders.push(loader)
    return loader
  }
}
