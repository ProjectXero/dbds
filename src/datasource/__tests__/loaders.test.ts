import { LoaderFactory } from '../loaders'
import { GetDataMultiFunction } from '../loaders/types'
import { match } from '../loaders/utils'

interface DummyRowType {
  id: number
  name: string
  code: string
}

const columnTypes: Record<keyof DummyRowType, string> = {
  id: 'anything',
  name: 'anything',
  code: 'anything',
}

describe(LoaderFactory, () => {
  const dummyBatchFn = async (): Promise<DummyRowType[]> => {
    return [
      { id: 1, name: 'aaa', code: 'abc' },
      { id: 2, name: 'bbb', code: 'def' },
      { id: 3, name: 'Aaa', code: 'ghi' },
      { id: 4, name: 'Bbb', code: 'ABC' },
      { id: 5, name: 'AAa', code: 'DEF' },
      { id: 6, name: 'BBb', code: 'GHI' },
      { id: 7, name: 'AAA', code: 'abc' },
      { id: 8, name: 'BBB', code: 'def' },
      { id: 9, name: 'zzz', code: 'ghi' },
      { id: 10, name: 'ccc', code: 'ABC' },
      { id: 11, name: 'Ccc', code: 'DEF' },
      { id: 12, name: 'CCc', code: 'GHI' },
      { id: 13, name: 'CCC', code: 'zzz' },
    ]
  }

  const factory = new LoaderFactory<DummyRowType>(dummyBatchFn, dummyBatchFn, {
    columnTypes,
  })

  describe('multi: false, ignoreCase: false', () => {
    const loader = factory.create('id', 'any', {
      multi: false,
      ignoreCase: false,
    })

    it('returns the first matching result', async () => {
      expect(await loader.load(1)).toMatchSnapshot()
    })

    it('is able to load many items', async () => {
      expect(await loader.loadMany([1, 2, 3])).toMatchSnapshot()
    })

    it('returns undefined with no matching results', async () => {
      expect(await loader.load(-Infinity)).toBeUndefined()
    })
  })

  describe('multi: true, ignoreCase: false', () => {
    const loader = factory.create('code', 'any', {
      multi: true,
      ignoreCase: false,
    })

    it('gets all matching results', async () => {
      expect(await loader.load('abc')).toMatchSnapshot()
    })

    it('is able to load many sets of items', async () => {
      expect(await loader.loadMany(['abc', 'def'])).toMatchSnapshot()
    })

    it('returns an empty array with no matching results', async () => {
      expect(await loader.load('any value with no match')).toHaveLength(0)
    })
  })

  describe('multi: false, ignoreCase: true', () => {
    const loader = factory.create('name', 'any', {
      multi: false,
      ignoreCase: true,
    })

    it('returns the first matching result', async () => {
      expect(await loader.load('AAA')).toMatchSnapshot()
    })

    it('is able to load many items', async () => {
      expect(await loader.loadMany(['AAA', 'BBB'])).toMatchSnapshot()
    })

    it('returns undefined with no matching results', async () => {
      expect(await loader.load('any value with no match')).toBeUndefined()
    })
  })

  describe('multi: true, ignoreCase: true', () => {
    const loader = factory.create('name', 'any', {
      multi: true,
      ignoreCase: true,
    })

    it('gets all matching results', async () => {
      expect(await loader.load('aaa')).toMatchSnapshot()
    })

    it('is able to load many sets of items', async () => {
      expect(await loader.loadMany(['aaa', 'bbb'])).toMatchSnapshot()
    })

    it('returns an empty array with no matching results', async () => {
      expect(await loader.load('any value with no match')).toHaveLength(0)
    })
  })

  describe('with no column type specified', () => {
    it('can accept options as the second parameter', async () => {
      const loader = factory.create('name', { multi: true, ignoreCase: true })
      expect(await loader.load('AAA')).toMatchSnapshot()
    })
  })

  describe('create', () => {
    it('can accept a getData override', async () => {
      const dummyRow = { id: 999, name: 'zzz', code: 'zzz' }
      const loader = factory.create('name', {
        getData: () => [dummyRow],
      })
      expect(await loader.load('zzz')).toMatchObject(dummyRow)
    })
  })

  describe('createMulti', () => {
    it('can create loaders with multiple columns', async () => {
      const loader = factory.createMulti(['name', 'code'], {
        multi: true,
        ignoreCase: true,
      })
      expect(await loader.load({ name: 'ccc', code: 'abc' })).toMatchSnapshot()
    })

    it('properly batches queries', async () => {
      const loader = factory.createMulti(['name', 'code'], {
        multi: true,
        ignoreCase: true,
      })
      expect(
        await Promise.all([
          loader.load({ name: 'ccc', code: 'abc' }),
          loader.load({ name: 'aaa', code: 'def' }),
        ])
      ).toMatchSnapshot()
    })

    it('properly caches queries', async () => {
      const dummyRow = { id: 999, name: 'zzz', code: 'zzz' }
      const getData = jest.fn(() => [dummyRow])
      const loader = factory.createMulti(['name', 'code'], { getData })
      await loader.load({ name: 'zzz', code: 'zzz' })
      await loader.load({ name: 'zzz', code: 'zzz' })
      expect(getData).toBeCalledTimes(1)
    })

    it('expects columnTypes of the same length if given', () => {
      expect(() => factory.createMulti(['name', 'code'])).not.toThrowError()

      expect(() =>
        factory.createMulti(['name', 'code'], ['onlyType1'])
      ).toThrowErrorMatchingInlineSnapshot(
        `"Same number of types and keys must be provided"`
      )

      expect(() =>
        factory.createMulti(['name', 'code'], ['type1', 'type2'])
      ).not.toThrowError()

      expect(() =>
        factory.createMulti(['name', 'code'], ['type1', 'type2', 'type3?????'])
      ).toThrowErrorMatchingInlineSnapshot(
        `"Same number of types and keys must be provided"`
      )
    })

    it('passes the given types correctly', async () => {
      const dummyRow = { id: 999, name: 'zzz', code: 'zzz' }
      const getData = jest.fn(
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        ((_args, _columns, _types, _options) => [
          dummyRow,
        ]) as GetDataMultiFunction<DummyRowType>
      )
      const loader = factory.createMulti(['name', 'code'], ['type1', 'type2'], {
        getData,
      })
      await loader.load({ name: 'zzz', code: 'zzz' })
      await loader.load({ name: 'zzz', code: 'zzz' })
      expect(getData.mock.calls[0][2]).toMatchInlineSnapshot(`
        Array [
          "type1",
          "type2",
        ]
      `)
    })
  })
})

describe(match, () => {
  it('exactly matches numbers', () => {
    expect(match(1, 1)).toBe(true)
    expect(match(1, 2)).toBe(false)
    expect(match(1.0, 1)).toBe(true)
    expect(match(1.001, 1)).toBe(false)
    expect(match(Infinity, 0)).toBe(false)
  })

  it('exactly matches strings', () => {
    expect(match('', '')).toBe(true)
    expect(match('a', '')).toBe(false)
    expect(match('a', 'a')).toBe(true)
    expect(match('', 'a')).toBe(false)
    expect(match('a', 'asdf')).toBe(false)
  })

  it('performs case-insensitive comparison on strings', () => {
    expect(match('a', 'A', true)).toBe(true)
    expect(match('a', 'a', true)).toBe(true)
    expect(match('', '', true)).toBe(true)
    expect(match('a', 'asdf', true)).toBe(false)
    expect(match('a', 'AA', true)).toBe(false)
  })
})
