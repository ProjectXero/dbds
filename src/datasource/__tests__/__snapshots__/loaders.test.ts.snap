// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoaderFactory createMulti can create loaders with multiple columns 1`] = `
Array [
  Object {
    "code": "ABC",
    "id": 10,
    "name": "ccc",
  },
]
`;

exports[`LoaderFactory createMulti properly batches queries 1`] = `
Array [
  Array [
    Object {
      "code": "ABC",
      "id": 10,
      "name": "ccc",
    },
  ],
  Array [
    Object {
      "code": "DEF",
      "id": 5,
      "name": "AAa",
    },
  ],
]
`;

exports[`LoaderFactory multi: false, ignoreCase: false is able to load many items 1`] = `
Array [
  Object {
    "code": "abc",
    "id": 1,
    "name": "aaa",
  },
  Object {
    "code": "def",
    "id": 2,
    "name": "bbb",
  },
  Object {
    "code": "ghi",
    "id": 3,
    "name": "Aaa",
  },
]
`;

exports[`LoaderFactory multi: false, ignoreCase: false returns the first matching result 1`] = `
Object {
  "code": "abc",
  "id": 1,
  "name": "aaa",
}
`;

exports[`LoaderFactory multi: false, ignoreCase: true is able to load many items 1`] = `
Array [
  Object {
    "code": "abc",
    "id": 1,
    "name": "aaa",
  },
  Object {
    "code": "def",
    "id": 2,
    "name": "bbb",
  },
]
`;

exports[`LoaderFactory multi: false, ignoreCase: true returns the first matching result 1`] = `
Object {
  "code": "abc",
  "id": 1,
  "name": "aaa",
}
`;

exports[`LoaderFactory multi: true, ignoreCase: false gets all matching results 1`] = `
Array [
  Object {
    "code": "abc",
    "id": 1,
    "name": "aaa",
  },
  Object {
    "code": "abc",
    "id": 7,
    "name": "AAA",
  },
]
`;

exports[`LoaderFactory multi: true, ignoreCase: false is able to load many sets of items 1`] = `
Array [
  Array [
    Object {
      "code": "abc",
      "id": 1,
      "name": "aaa",
    },
    Object {
      "code": "abc",
      "id": 7,
      "name": "AAA",
    },
  ],
  Array [
    Object {
      "code": "def",
      "id": 2,
      "name": "bbb",
    },
    Object {
      "code": "def",
      "id": 8,
      "name": "BBB",
    },
  ],
]
`;

exports[`LoaderFactory multi: true, ignoreCase: true gets all matching results 1`] = `
Array [
  Object {
    "code": "abc",
    "id": 1,
    "name": "aaa",
  },
  Object {
    "code": "ghi",
    "id": 3,
    "name": "Aaa",
  },
  Object {
    "code": "DEF",
    "id": 5,
    "name": "AAa",
  },
  Object {
    "code": "abc",
    "id": 7,
    "name": "AAA",
  },
]
`;

exports[`LoaderFactory multi: true, ignoreCase: true is able to load many sets of items 1`] = `
Array [
  Array [
    Object {
      "code": "abc",
      "id": 1,
      "name": "aaa",
    },
    Object {
      "code": "ghi",
      "id": 3,
      "name": "Aaa",
    },
    Object {
      "code": "DEF",
      "id": 5,
      "name": "AAa",
    },
    Object {
      "code": "abc",
      "id": 7,
      "name": "AAA",
    },
  ],
  Array [
    Object {
      "code": "def",
      "id": 2,
      "name": "bbb",
    },
    Object {
      "code": "ABC",
      "id": 4,
      "name": "Bbb",
    },
    Object {
      "code": "GHI",
      "id": 6,
      "name": "BBb",
    },
    Object {
      "code": "def",
      "id": 8,
      "name": "BBB",
    },
  ],
]
`;

exports[`LoaderFactory with no column type specified can accept options as the second parameter 1`] = `
Array [
  Object {
    "code": "abc",
    "id": 1,
    "name": "aaa",
  },
  Object {
    "code": "ghi",
    "id": 3,
    "name": "Aaa",
  },
  Object {
    "code": "DEF",
    "id": 5,
    "name": "AAa",
  },
  Object {
    "code": "abc",
    "id": 7,
    "name": "AAA",
  },
]
`;
